bsip-0040.md

```diff
@@ -0,0 +1,170 @@
+ - Cold Storage Key: Only allow to move funds to the Hot Wallet
+ 
+The above list of named keys is nothing that is known to the backend as the backend should have an abstract implementation.
+ The UI could provide a button "Create Trading Key" that properly configures the respective custom active permission entry.
+
+# Rational
+
+Custom active permission is a list of `custom active authorities`. A `custom active authority` contains an `operation_id`, an `authority` (just like with active permission) and `asserts` than can be used to restrict arguments and is only valid a certain time period (`valid_from` and `valid_to`). When handling incoming signed transactions, the backend checks for each operation if there is a `custom active authority` for any of the required accounts. If all operations within the transaction have a matching `custom active authority` (match means same `operationid`, now is within `valid_to` and `valid_from` and all `asserts` pass), then active authority of the corresponding account is granted. 
+
+# Specification
+
+All descriptions in this section are on a pseudo/prosa level and no recommendation how it can best be implemented or serialized. They are meant to facilitate the understanding.
+
+### Custom active permission and custom active authority
+
+A `custom_active_permission` looks like follows (in JSON/pseudo for clarification):
+```
+custom_active_permission = list of custom_active_authority items
+custom_active_authority = {
+    valid_from,    // timestamp when this is active, defaults to now
+    valid_to,      // timestamp when this is invalid, defaults to 1 month
+    operationid,   // operationid of the target operation,
+    authority,     // same as for the existing authortities (multisig with weighted accounts or keys),
+    asserts        // see below
+}
+```
+Note: This assumes `custom_active_permission` is stored within `account_object`. Actual implementation details left to the implementer, as long as every `custom_active_permission` can be assigned to exactly one account.
+
+### Asserts
+
+The `asserts` is a list of `assert_objects` that are all together evaluated with `and` logic to evaluate a match
+```
+asserts = list of assert_objects
+assert_object = {
+    argument, // target argument of the operationid
+    function, // functionid to do the assert 
+    data,     // stores data specific to the chosen function
+    state     // if this assert is statefull
+}
+```
+List of possible asserts are:
+
+| function        | data           | state  |
+| ------------- |:-------------:| -----:|
+| `any`      | [`list`, `of`, `allowed`, `values`] | stateless |
+| `range`      | [`min`, `max`] | stateless |
+| `length`      | [`min`, `max`] | stateless |
+| `limit` | [`max_cumsum`, `interval_in_sec`]      | [`current_cumsum`, `interval_began`] |
+| `limit_monthly` | [`max_cumsum`, `interval_in_months`]      | [`current_cumsum`, `interval_began`] |
```

Comment:
I have added a `contains_only` assert, which flips your logic around. A custom active authority is more likely to allow just one specific attribute rather than everything except one attribute.

Reply:
As I've stated, this part is hard.

1. fields need to be standardized, otherwise the back end won't understand. That said, we need to define mapping rules for fields of each operation, primarily due to existence of nested structures, e.g. `price` or `asset_options`.
* `price` is a 2-level struct with 4 data fields: `{"base":{"amount":1,"asset_id":"1.3.1"},"quote":{"amount":2,"asset_id":"1.3.2"}}`. 
  * by the way, usually a `price` need to be validated as a whole, e.g. we need to assert `base.amount/quote.amount < x`.
* `asset_options` is a nested struct as well, it even has a `price` object inside.

  Actually, we need to define format of "argument" as well. When the field type is `static_variant`, it would be hard to define.

2. An operation have to contain all the needed fields. My example above is that the values in the incoming transaction need to be same as current values on chain. But when creating the asserts (rules) we don't know the exact value or don't care what the value is when processing a transaction, so can't put a value there.

  A "contains_only" assert makes some sense, but it's operation-specific:
* for operations like "account_update" or "asset_update" we need to check whether the values are same as current on-chain values;
* for operations like "call_order_update" or "account_whitelist" we need to check whether the values are zero or empty, because they're deltas;
* for operations like "transfer" perhaps we need to check whether the memo is empty if it's not in the list;
* for operations like "limit_order_create" it seems there is nothing to check so far.

Last, not like other functions which each applies to an argument/field, this "contains_only" function applies to the whole operation?

Reply:
1. The `lt` asserts (and similar) have now an explicit type conversion for `price`, others may be added as well. This might be unnecessary considering the new `attribute_assert`. I have also included a line that `argument` type can be added to `assert objects` if necessary. I want to avoid operation specific logic in the `custom active authority` check at all costs.

2. On-chain state checking should not be possible (would mean that we need to know which operation store what and where on chain). In general: If there is no assert defined, any value is allowed.

3. contains_only does no longer exist

Reply:
I discussed this with Stefan already. Nested objects are probably out of scope here.
In case of voting (as an example), we would propose a new operation that can only vote.
I suggest we do the same with core exchange rate ..